pkg/cache/v3/delta_test.go

```diff
@@ -0,0 +1,148 @@
+			}
+		})
+	}
+
+	for _, typ := range testTypes {
+		watches[typ], _ = c.CreateDeltaWatch(&discovery.DeltaDiscoveryRequest{
+			Node: &core.Node{
+				Id: "node",
+			},
+			TypeUrl:                typ,
+			ResourceNamesSubscribe: names[typ],
+		}, &stream.StreamState{IsWildcard: false, ResourceVersions: vm[typ]})
+	}
+
+	if count := c.GetStatusInfo(key).GetNumDeltaWatches(); count != len(testTypes) {
+		t.Errorf("watches should be created for the latest version, saw %d watches expected %d", count, len(testTypes))
+	}
+
+	// set partially-versioned snapshot
+	snapshot2 := snapshot
+	snapshot2.Resources[types.Endpoint] = cache.NewResources(version2, []types.Resource{resource.MakeEndpoint(clusterName, 9090)})
+	if err := c.SetSnapshot(key, snapshot2); err != nil {
+		t.Fatal(err)
+	}
+	if count := c.GetStatusInfo(key).GetNumDeltaWatches(); count != len(testTypes)-1 {
+		t.Errorf("watches should be preserved for all but one: %d", count)
+	}
+
+	// validate response for endpoints
+	select {
+	case out := <-watches[testTypes[0]]:
+		if !reflect.DeepEqual(cache.IndexRawResourcesByName(out.(*cache.RawDeltaResponse).Resources), snapshot2.GetResources(rsrc.EndpointType)) {
+			t.Fatalf("got resources %v, want %v", out.(*cache.RawDeltaResponse).Resources, snapshot2.GetResources(rsrc.EndpointType))
+		}
+		vMap := out.GetNextVersionMap()
+		vm[testTypes[0]] = vMap
+	case <-time.After(time.Second):
+		t.Fatal("failed to receive snapshot response")
+	}
+}
+
+func TestConcurrentSetDeltaWatch(t *testing.T) {
+	c := cache.NewSnapshotCache(false, group{}, logger{t: t})
+	for i := 0; i < 50; i++ {
+		version := fmt.Sprintf("v%d", i)
+		func(i int) {
+			t.Run(fmt.Sprintf("worker%d", i), func(t *testing.T) {
+				t.Parallel()
+				id := fmt.Sprintf("%d", i%2)
+				var cancel func()
+				if i < 25 {
+					snap := cache.Snapshot{}
+					snap.Resources[types.Endpoint] = cache.NewResources(version, []types.Resource{resource.MakeEndpoint(clusterName, uint32(i))})
+					c.SetSnapshot(id, snap)
+				} else {
+					if cancel != nil {
+						cancel()
+					}
+
+					_, cancel = c.CreateDeltaWatch(&discovery.DeltaDiscoveryRequest{
+						Node: &core.Node{
+							Id: id,
+						},
+						TypeUrl:                rsrc.EndpointType,
+						ResourceNamesSubscribe: []string{clusterName},
+					}, &stream.StreamState{IsWildcard: true, ResourceVersions: nil})
+				}
+			})
+		}(i)
+	}
+}
+
+func TestSnapshotCacheDeltaWatchCancel(t *testing.T) {
+	c := cache.NewSnapshotCache(true, group{}, logger{t: t})
+	for _, typ := range testTypes {
+		_, cancel := c.CreateDeltaWatch(&discovery.DeltaDiscoveryRequest{
+			Node: &core.Node{
+				Id: key,
+			},
+			TypeUrl:                typ,
+			ResourceNamesSubscribe: names[typ],
+		}, &stream.StreamState{IsWildcard: true, ResourceVersions: nil})
+
+		// Cancel the watch
+		cancel()
+	}
+	// should be status info for the node
+	if keys := c.GetStatusKeys(); len(keys) == 0 {
+		t.Error("got 0, want status info for the node")
```

Comment:
this error message could be more useful, maybe `expected to see a status info registered for watch, saw %d entries`

Reply:
Yup agreed, will change that.